### PR TITLE
Handle empty timeseries input to GordoDataset

### LINKED ISSUE #686: join_timeseries: handle empty timeseries
Currently we crash with 
```
  File "/usr/local/lib/python3.6/site-packages/gordo_components/dataset/base.py", line 77, in join_timeseries
    tz=series.index[0].tzinfo
  File "/usr/local/lib/python3.6/site-packages/pandas/core/indexes/datetimes.py", line 1220, in __getitem__
    result = self._data.__getitem__(key)
  File "/usr/local/lib/python3.6/site-packages/pandas/core/arrays/datetimelike.py", line 426, in __getitem__
    val = getitem(key)
IndexError: index 0 is out of bounds for axis 0 with size 0
```
https://github.com/equinor/gordo-components/blob/cc942329f58d0721216267d158ca5d52879f9cf8/gordo_components/dataset/base.py#L77

Maybe we should either raise a better error, or handle it so we return an empty dataframe (which is maybe the right thing to do?). If we end up doing the latter we should probably make sure we have some tests somewhere so we dont end up training models on no data.

### PR DESCRIPTION
**Pull request by ryanjdillon**
This PR does the following:
* Converts `tests/gordo_components/dataset/test_dataset.py` to pytest format
* Raises a new `InsufficientDataError` if not enough row in a sensor `Series`.

closes #686

### CODE REVIEW

File: gordo_components/dataset/base.py [hunk 1]
```diff
@@ -73,6 +74,12 @@ def join_timeseries(
         resampled_series = []
 
         for series in series_iterable:
+            if not series.size:
+                dt_index = pd.date_range(
+                    start=resampling_startpoint, end=resampling_endpoint, freq="24H"
+                )
+                series = pd.Series([np.NaN] * len(dt_index), index=dt_index, tz=UTC)
```
Comment (epa095) [inline gordo_components/dataset/base.py]: Nha, it needs to be empty instead. Note as the documentation to this function says, it is guaranteed to return a dataframe without NaNs.

File: gordo_components/dataset/base.py [hunk 2]
```diff
@@ -73,6 +74,12 @@ def join_timeseries(
         resampled_series = []
 
         for series in series_iterable:
+            if not series.size:
```
Comment (epa095) [inline gordo_components/dataset/base.py]: I guess you rely on 0 beeing false here? I must admit I prefer explicit check like `==0` if it is guaranteed to be 0. Or can it sometimes be None?

**Review by epa095 — COMMENTED** [review]
good start :-)

File: gordo_components/dataset/base.py [hunk 3]
```diff
@@ -70,78 +79,87 @@ def join_timeseries(
...
+                    f"Series {repr(series.name)} size ({series.size}) is less than or equal to the minimum required ({row_threshold})."
                 )
 
-            elif series.index[0] < resampling_startpoint:
-                msg = (
-                    f"Error - for {series.name}, first timestamp "
-                    f"{series.index[0]} is before the resampling start point "
-                    f"{startpoint_sametz}"
+            # Resample if we have a resolution set, otherwise simply join the series.
+            if bool(resolution):
```
Comment (milesgranger) [inline gordo_components/dataset/base.py]: A small thing, but what case are we trying to cover by explicitly calling `bool` here?

File: gordo_components/dataset/base.py [hunk 4]
```diff
@@ -70,78 +79,87 @@ def join_timeseries(
...
+                    resampled.columns = pd.MultiIndex.from_product(
+                        [[series.name], resampled.columns],
+                        names=["tag", "aggregation_method"],
+                    )
+                filled = resampled.fillna(method="ffill")
+                _series_iterable.append(filled)
+            else:
+                _series_iterable.append(series)
+
+        data = pd.concat(_series_iterable.__iter__(), axis=1, join="inner")
```
[the remark this answers is not in the document]
Comment (ryanjdillon) [reply gordo_components/dataset/base.py]: :+1:  That's a space-cadet-residual I missed after debugging something.

File: gordo_components/dataset/datasets.py [hunk 5]
```diff
@@ -155,17 +155,13 @@ def get_data(self) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
...
-            data = self.join_timeseries(
-                series_iter,
-                self.from_ts,
-                self.to_ts,
-                self.resolution,
-                aggregation_methods=self.aggregation_methods,
-            )
-        else:
-            data = pd.concat(series_iter, axis=1, join="inner")
+        data = self.join_timeseries(
```
[the remark this answers is not in the document]
Comment (ryanjdillon) [reply gordo_components/dataset/datasets.py]: I personally like this in that it keeps all of the series join logic in one method. The nan dropping I can move within the `if resolution` block.

File: gordo_components/dataset/base.py (hunk 3 above)
Comment (ryanjdillon) [reply gordo_components/dataset/base.py]: :+1:  That's another residual from debugging this weird error I was getting.

File: gordo_components/dataset/base.py (hunk 2 above)
Comment (ryanjdillon) [reply gordo_components/dataset/base.py]: After some discussion with Miles, I'm now catching the index error in `base.GordoDataset.get_data`, so this has been removed.

File: gordo_components/dataset/datasets.py [hunk 6]
```diff
@@ -61,6 +65,7 @@ def __init__(
         row_filter_buffer_size: int = 0,
         asset: Optional[str] = None,
         default_asset: Optional[str] = None,
+        row_threshold: int = 0,
```
Comment (milesgranger) [inline gordo_components/dataset/datasets.py]: Could we rename this to something like `n_samples_threshold` or something?

File: gordo_components/dataset/datasets.py [hunk 7]
```diff
@@ -157,13 +165,23 @@ def get_data(self) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
...
+                    self.from_ts,
+                    self.to_ts,
+                    self.resolution,
+                    aggregation_methods=self.aggregation_methods,
+                )
+                if len(data) <= self.row_threshold:
+                    raise InsufficientDataError(
+                        f"The length of the generated DataFrame ({len(data)}) does not exceed the "
+                        f"specified required threshold for number of rows ({self.row_threshold})."
+                    )
```
Comment (milesgranger) [inline gordo_components/dataset/datasets.py]: I was thinking rather this should be outside of here and after the possible joining without resolution. That way both are checked.

**Review by milesgranger — COMMENTED** [review]
I like this approach better then the previous work. :+1:

File: gordo_components/dataset/datasets.py [hunk 8]
```diff
@@ -21,6 +21,10 @@
 logger = logging.getLogger(__name__)
 
 
+class InsufficientDataError(ValueError):
```
Comment (epa095) [inline gordo_components/dataset/datasets.py:24]: Also remember to add it to this handy map: https://github.com/equinor/gordo-components/blob/50d5a9e96550b2c9c4eda02bcc07c6f725da7cca/gordo_components/cli/cli.py#L31

Comment (ryanjdillon) [reply gordo_components/dataset/datasets.py:24]: Oooh... right.

File: gordo_components/dataset/datasets.py (hunk 7 above)
Comment (ryanjdillon) [reply gordo_components/dataset/datasets.py]: OK, got it. See what you think now :+1:

**Review by milesgranger — APPROVED** [review]
Nice! :clap:  Not sure if @epa095 has any extra comments or not though.

**Review by epa095 — APPROVED** [review]
LGTM

**Pull request merged**